cmd/bnsd/scenarios/bnsdtest/bnsdtest.go

```diff
@@ -22,114 +25,150 @@ import (
-	logger.Info("Loaded Alice key", "addressID", alice.PublicKey().Address().String(), "derivationPath", *derivationPath)
+func startRemoteBnsd(t testing.TB, env *EnvConf) (cleanup func()) {
+	env.Client = client.NewClient(client.NewHTTPConnection(*tendermintAddress))
 
-	if *tendermintAddress != testLocalAddress {
-		bnsClient = client.NewClient(client.NewHTTPConnection(*tendermintAddress))
-		var err error
-		chainID, err = bnsClient.ChainID()
-		if err != nil {
-			logger.Error("Failed to fetch chain id", "cause", err)
-			os.Exit(1)
-		}
-		rpcAddress = *tendermintAddress
-		os.Exit(m.Run())
+	if chainID, err := env.Client.ChainID(); err != nil {
+		t.Fatalf("failed to fetch chain id: %s", err)
+	} else {
+		env.ChainID = chainID
 	}
 
-	config := rpctest.GetConfig()
-	config.Moniker = "SetInTestMain"
-	chainID = config.ChainID()
+	env.RpcAddress = *tendermintAddress
+	return func() {}
+}
+
+func startLocalBnsd(t testing.TB, env *EnvConf) (cleanup func()) {
+	tmWorkDir := fmt.Sprintf("bnsd_%s_%d", t.Name(), time.Now().UnixNano())
+	tmConf := buildConfig(t, tmWorkDir)
+
+	tmConf.Moniker = "SetInTestMain"
+	env.ChainID = tmConf.ChainID()
+
+	env.RpcAddress = "http://localhost" + tmConf.RPC.ListenAddress[strings.LastIndex(tmConf.RPC.ListenAddress, ":"):]
 
-	rpcAddress = "http://localhost" + config.RPC.ListenAddress[strings.LastIndex(config.RPC.ListenAddress, ":"):]
-	app, err := initApp(config, alice.PublicKey().Address())
+	initGenesis(t, env, tmConf.GenesisFile())
+
+	bnsd, err := app.GenerateApp(&server.Options{
+		MinFee: env.MinFee,
+		Home:   tmConf.RootDir,
+		Logger: env.Logger,
+		Debug:  false,
+	})
 	if err != nil {
-		logger.Error("Failed to init app", "cause", err)
-		os.Exit(1)
+		t.Fatalf("cannot generate application: %s", err)
 	}
 
-	code := weaveClient.TestWithTendermint(app, func(n *nm.Node) {
-		node = n
-		bnsClient = client.NewClient(client.NewLocalConnection(node))
-	}, m)
-	os.Exit(code)
+	env.Node = newTendermint(t, tmConf, bnsd)
+	if err := env.Node.Start(); err != nil {
+		t.Fatalf("cannot start tendermint node: %s", err)
+	}
 
-}
+	waitForRPC(t, tmConf)
+	waitForGRPC(t, tmConf)
 
-func delayForRateLimits() {
-	time.Sleep(*delay)
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	if _, err := weaveClient.NewLocalClient(env.Node).WaitForNextBlock(ctx); err != nil {
+		t.Fatalf("cannot start tendermint application: %s", err)
+	}
+
+	env.Client = client.NewClient(client.NewLocalConnection(env.Node))
+
+	return func() {
+		env.Node.Stop()
+		env.Node.Wait()
+		os.RemoveAll(tmWorkDir)
+	}
 }
 
-func initApp(config *cfg.Config, addr weave.Address) (abci.Application, error) {
-	opts := &server.Options{
-		MinFee: coin.Coin{},
-		Home:   config.RootDir,
-		Logger: logger,
-		Debug:  false,
+const TendermintLocalAddr = "localhost:46657"
+
+var (
+	tendermintAddress = flag.String("address", TendermintLocalAddr, "destination address of tendermint rpc")
```

Comment:
Don't really like this flag value being accessed directly above in StartBnsd.
We can parse the flags in some setup, but then should pass this stuff into StartBnsd (along with MinFee)

Reply:
Do you have anything concrete in mind?

I think there is no way around using flag. It is parsed once and it is global as it never changes. I was thinking of pulling it to a separate file, maybe in a variable scope to make it look nicer.

Reply:
This level of parsing is fine.
I just didn't like `StartBnsd` referencing it directly. So there is no way for some testcase to override it.